### PR TITLE
Allow unbounded liquidity when creating new market

### LINKED ISSUE #478: Creating a pool with large liquidity
Deploying a swap pool with a large amount of liquidity seems super complicated right now. If I want more than 150 ZTG, I get the `MaxInRatio` error, so if I want to deploy a pool with 500 ZTG, then I have to call `buy_complete_set`, call `deploy_swap_pool` and then `pool_join` a couple of times, each time with 1.5 times the amount of the previous call.

Is there another way to do this? If not, why not make amount, which is currently hardcoded to `MinLiquidity`, a parameter of `deploy_swap_pool`?

### LINKED ISSUE #492: [swaps] Deploying swap pool with uneven `amount_outcome_assets` can raise a `MaxInRatio` error
If you run the following transaction, you'll get a `MaxInRatio` error:

```js
api.tx.predictionMarkets.createCpmmMarketAndDeployAssets(
  ALICE,
  api.createType("MarketPeriod", { Block: [0, 100] }),
  api.createType("MultiHash", { Sha3_384: METADATA }),
  api.createType("MarketType", { Categorical: 2 }),
  api.createType("MarketDisputeMechanism", "SimpleDisputes"),
  100 * BASE,
  [100 * BASE, 10000 * BASE],
  [MIN_WEIGHT, MIN_WEIGHT, MIN_WEIGHT],
  [0, 0]
);
```

What happened? `deploy_swap_pool_and_additional_liquidity` buys 10000 complete sets, then sells 9900 of the first outcome asset back into the pool, which has a balance of 100. This is not allowed.

Even if the `MaxInRatio` error was not raised, there'd still be a conceptual problem with this design: The LP spends 10000 ZTG on complete sets, then sells almost half of the tokens they receive (9900 of the first outcome token) back to the pool, but can never expect to receive more than a little under 100 ZTG back (because there's no more liquidity in the pool). Even in more realistic scenarios, this makes it hard for the user to understand what amount of ZTG they are paying, and what they're getting out of it.

My first reaction to this is that we should maybe remove the `keep` parameter and always make the LP keep the minted outcome tokens.

### PR DESCRIPTION
**Pull request by maltekliemann**
Fixes #492, fixes #478.

### CODE REVIEW

**Comment by maltekliemann** [comment]
I need some advice for this one: `deploy_swap_pool_and_additional_liquidity` and `deploy_swap_pool_for_market` now only differ in one thing: The former buys complete sets before deploying, the latter doesn't. Are these function sufficiently different to keep both of them around? How should they be called?

Other than that, you'll notice that I've removed the feature which allows the pool deployer to add different amounts of tokens to the pool. This feature doesn't make a lot of sense to me, especially given that it's implemented by using single-asset joins which will fail if differences between asset amounts are too large.

**Comment by maltekliemann** [comment]
This also fixes #268.

File: zrml/swaps/src/lib.rs
Comment (Chralt98) [inline zrml/swaps/src/lib.rs:1365]:
```suggestion
                ensure!(amount.is_some(), Error::<T>::InvalidAmountArgument);
```

**Review by Chralt98 — COMMENTED** [review]
This will simplify the deployment of a pool, because it's now allowed only to use an equal amount of outcome assets.

File: primitives/src/traits/swaps.rs
Comment (maltekliemann) [reply primitives/src/traits/swaps.rs:33]: :thumbsup: Mitigated in 61c038abc5a3a1ee93c483aa894b259fd7e2e391

File: zrml/prediction-markets/src/tests.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:1916]: The market has five outcome assets, but the pool must contain six assets: The outcome assets and the base asset. I'll leave an explainatory comment.

Mitigrated in 61c038abc5a3a1ee93c483aa894b259fd7e2e391

File: zrml/prediction-markets/src/tests.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:2063]: Mitigated in 61c038abc5a3a1ee93c483aa894b259fd7e2e391

File: zrml/swaps/src/lib.rs
Comment (maltekliemann) [reply zrml/swaps/src/lib.rs:1362]: `unwrap_or_else` just requires a `[FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html)() -> Balance::<T>`. `BalanceOf::<T>::zero`, as a function pointer, satisfies that interface. No need to wrap it in a closure.

Performance-wise, both get optimized to the same thing when building release, as you can test here: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=4d0905106730cf6255349c58a2e6b954

File: zrml/prediction-markets/src/tests.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:1350]: Just a random weight.

File: zrml/swaps/src/benchmarks.rs
Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs:318]: Only explanation I have is that this is a typo.

File: zrml/swaps/src/benchmarks.rs
Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs:62]: The benchmark for `pool_exit` removes `100 * BASE` from the pool. It hasn't yet been fully established what happens to a pool if any of the balances is reduced to zero. So I'm just using a larger number.

**Comment by Chralt98** [comment]
> Are these function sufficiently different to keep both of them around? How should they be called?

`deploy_swap_pool_and_additional_liquidity` is essentially a lazy extrinsic for the usual use of `buy_complete_set` and then  `deploy_swap_pool_for_market`. It now breaks down to the user usage of polkadot-js. I mean to code the calls of two extrinsics in a specific order is not the problem. 

> /// The sender must have enough funds to cover all of the required shares to seed the pool.

So this doc comment explains why there is `deploy_swap_pool_and_additional_liquidity`. It reduces the amount of extrinsic calls on the polkadot-js site. I don't see any reason to remove this extrinsic, when it provides an usual action.

File: zrml/swaps/src/benchmarks.rs
Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs:62]: But I realize that it's a strange default value. Fixing this...

File: zrml/swaps/src/benchmarks.rs
Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs:323]: That was to make sure that we move away from the `LimitIn` error that would be caused by a small `max_assets_in`, but it's not necessary. I'll remove this.

Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs]: 0e0360ea13278fe5a045e3fc24fa4becd3b1c1b6

File: zrml/swaps/src/benchmarks.rs
Comment (maltekliemann) [reply zrml/swaps/src/benchmarks.rs]: 0e0360ea13278fe5a045e3fc24fa4becd3b1c1b6

**Review by Chralt98 — APPROVED** [review]
Looks good to me.

File: zrml/prediction-markets/src/lib.rs
Comment (sea212) [inline zrml/prediction-markets/src/lib.rs:584]: I think we don't need those reads anymore, because the two places that did storage reads were removed (all other reads are contained within functions which are benchmarked).
```suggestion
```

File: zrml/prediction-markets/src/lib.rs
Comment (sea212) [inline zrml/prediction-markets/src/lib.rs:567]: Technically speaking this function is not necessary anymore, because anyone can use `pallet-utility` to batch the `buy_complete_set` and `deploy_swap_pool_for_market` functions into one transaction. However, I am fine with keeping it since it's more straightforward and we already have it.

File: zrml/swaps/src/lib.rs
Comment (sea212) [inline zrml/swaps/src/lib.rs:813]: Let's maintain the lexicographical ordering within the event enum.

File: zrml/prediction-markets/src/tests.rs
Comment (sea212) [inline zrml/prediction-markets/src/tests.rs:1350]: This test fails once `MinWeight` changes to something greater than `2 * BASE` in primitives crate.

**Review by sea212 — COMMENTED** [review]
Just some minor things, looks good overall.
I remember the first iteration of the combined market creation and pool deployment function only allowed one amount, but then there was the request from the frontend team to implement the feature to supply multiple amounts. I see there is no transparent and fair way to implement this right now, `buy_complete_set` is limited to the same amounts. However, we should communicate that change and the reasoning behind that clearly.

File: zrml/prediction-markets/src/lib.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/lib.rs:567]: Let's keep it for now. We can remove it at some point down the road.

File: zrml/swaps/src/lib.rs
Comment (maltekliemann) [reply zrml/swaps/src/lib.rs]: :thumbsup:

File: zrml/prediction-markets/src/tests.rs
Comment (maltekliemann) [reply zrml/prediction-markets/src/tests.rs:1350]: So will any test that uses `create_initial_pool`. We could change `2 * BASE` to `MinWeight`, but you could still argue that the test will fail if we change the minimum of maximum number of assets in the pool below or above 5. Or the min/max category count.

Personally, I don't think it's necessary that we future proof our tests against any possible change of constants. I guess I could replace `MinWeight` with a mock-specific weight. But if we start there, where do we stop?

File: primitives/src/traits/swaps.rs
Comment (sea212) [reply primitives/src/traits/swaps.rs:24]: Oh now I see it \*facepalm\* . I would find it easier to distinguish if there is an extra indentation, such that it's not aligned with the `parameter names`.

File: zrml/prediction-markets/src/tests.rs
Comment (sea212) [reply zrml/prediction-markets/src/tests.rs:1350]: I see, it's difficult to draw the line there. But it's also inconsistent in the current state, for example for `MinLiquidity` we use the constant instead of `1_000_000_000_000` that is supplied by the runtime. Generally I would use constants specified within the test runtime. The value `5` is arbitrary for that test and not a global constant within that pallet.

What I also realized is that it might be better for the reader to name the values instead of using raw numbers, in cases where numbers repeatedly used a constant is appropriate. But that's something for another meeting I guess - Now that we're plenty of people with different styles, I guess it would be good to have some guidelines that we try to adhere to.

**Comment by sea212** [comment]
I'll review and approve once the merge conflicts are resolved.

**Review by sea212 — APPROVED** [review]
LGTM. #268 is not closed by this, since the underlying question is not resolved. The issue requests to figure out if buying and deploying a complete set with different amounts, i.e. `[150, 120]` is possible, such that the liquidity provider has complete freedom over the amounts and is not forced to keep some outcome tokens to achieve that. I'll remove it from the list of issue to close with this PR.

**Comment by maltekliemann** [comment]
Good catch, thanks! :thumbsup:

Regarding the style suggestions (naming in tests, indentation in doc-strings): I'm making notes of all these, so that we can compile a concise style guide for the team.

**Pull request merged**
